### PR TITLE
BUG/TST: Add searchsorted tests

### PR DESCRIPTION
**Pull request by dsaxton**
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`

xref: https://github.com/pandas-dev/pandas/issues/32845

### CODE REVIEW

**Comment by jbrockmendel** [comment]
LGTM

File: pandas/tests/indexes/interval/test_interval.py [hunk 1]
```diff
@@ -863,3 +863,23 @@ def test_dir():
...
+def test_searchsorted_different_argument_classes(klass):
+    # https://github.com/pandas-dev/pandas/issues/32762
+    values = IntervalIndex([Interval(0, 1), Interval(1, 2)])
+    result = values.searchsorted(klass(values))
+    expected = np.array([0, 1], dtype=result.dtype)
+
+    tm.assert_numpy_array_equal(result, expected)
+
+
+@pytest.mark.parametrize(
```
Comment (jreback) [inline pandas/tests/indexes/interval/test_interval.py]: can you add a Timedelta  & Period test cases here as well (followon PR is ok)

Comment (jreback) [reply pandas/tests/indexes/interval/test_interval.py]: actually if you'd do in this one

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/arrays/datetimelike.py [hunk 2]
```diff
@@ -804,6 +804,16 @@ def searchsorted(self, value, side="left", sorter=None):
         indices : array of ints
             Array of insertion points with the same shape as `value`.
         """
+        if is_list_like(value) and not isinstance(value, type(self)):
```
Comment (dsaxton) [inline pandas/core/arrays/datetimelike.py]: The tests added for PeriodIndex were failing so had to include this bug fix

Comment (jbrockmendel) [reply pandas/core/arrays/datetimelike.py]: does this need to go before the str and _recognized_scalar checks below?  IIRC putting them in this order is a pattern in this and related modules

File: pandas/tests/indexes/period/test_tools.py [hunk 3]
```diff
@@ -64,6 +66,17 @@ def test_searchsorted(self, freq):
         with pytest.raises(IncompatibleFrequency, match=msg):
             pidx.searchsorted(Period("2014-01-01", freq="5D"))
 
+    @pytest.mark.parametrize("klass", [list, np.array, array, Series])
+    def test_searchsorted_different_argument_classes(self, klass):
+        pidx = PeriodIndex(
+            ["2014-01-01", "2014-01-02", "2014-01-03", "2014-01-04", "2014-01-05"],
+            freq="D",
+        )
+        result = pidx.searchsorted(klass(pidx))
```
Comment (jbrockmendel) [inline pandas/tests/indexes/period/test_tools.py]: can you repeat this with pidx._data.searchsorted

File: pandas/tests/indexes/interval/test_interval.py [hunk 4]
```diff
@@ -863,3 +863,23 @@ def test_dir():
     index = IntervalIndex.from_arrays([0, 1], [1, 2])
     result = dir(index)
     assert "str" not in result
+
+
+@pytest.mark.parametrize("klass", [list, np.array, pd.array, pd.Series])
+def test_searchsorted_different_argument_classes(klass):
+    # https://github.com/pandas-dev/pandas/issues/32762
+    values = IntervalIndex([Interval(0, 1), Interval(1, 2)])
+    result = values.searchsorted(klass(values))
```
Comment (jbrockmendel) [inline pandas/tests/indexes/interval/test_interval.py]: can you repeat with values._data.searchsorted

File: pandas/core/arrays/datetimelike.py (hunk 2 above)
Comment (dsaxton) [reply pandas/core/arrays/datetimelike.py]: I don't think so, can move it down below

File: pandas/core/arrays/datetimelike.py [hunk 5]
```diff
@@ -846,14 +846,14 @@ def searchsorted(self, value, side="left", sorter=None):
...
+
             if not type(self)._is_recognized_dtype(value):
                 raise TypeError(
                     "searchsorted requires compatible dtype or scalar, "
                     f"not {type(value).__name__}"
                 )
-            value = type(self)(value)
-            self._check_compatible_with(value)
 
         if not (isinstance(value, (self._scalar_type, type(self))) or (value is NaT)):
```
Comment (jreback) [inline pandas/core/arrays/datetimelike.py]: is this line actually hit by anything? 858), this seems very duplicative

File: pandas/core/arrays/datetimelike.py [hunk 6]
```diff
@@ -846,14 +846,14 @@ def searchsorted(self, value, side="left", sorter=None):
         elif isinstance(value, self._recognized_scalars):
             value = self._scalar_type(value)
 
-        elif isinstance(value, np.ndarray):
+        if is_list_like(value) and not isinstance(value, type(self)):
```
Comment (jreback) [inline pandas/core/arrays/datetimelike.py]: can this be an elif?

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by dsaxton** [comment]
Carried over to https://github.com/pandas-dev/pandas/pull/32764 since there's overlap

**Pull request closed without merging**